### PR TITLE
East sussex salesforce

### PR DESCRIPTION
**Pull request by struan**

### CODE REVIEW

File: perllib/Integrations/SalesForceRest.pm [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+    return $service;
+}
+
+sub _get_pos {
+    my $s = shift;
+
+    my $bytes = decode_base64($s);
+    my $len = length($bytes) * 8;
+    my $pos_map = unpack("B$len", $bytes);
+    $pos_map =~ s/^(0*1).*$/$1/;
```
Comment (dracos) [inline perllib/Integrations/SalesForceRest.pm:244]: This function surely deserves an explanation ;)
It only cares of the length of the first string of zeroes?

File: perllib/Open311/Endpoint/Integration/SalesForceRest.pm [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+        my $defaults = $self->get_integration->config->{account_defaults};
+        my $mapping = $self->get_integration->config->{account_map};
+        my $map = { map { $mapping->{$_} => $args->{$_} } keys %$mapping };
+
+        my $account = {
+            %$defaults,
+            %$map
+        };
+
+        $id = $self->get_integration->create_object( 'Account', $account );
```
Comment (dracos) [inline perllib/Open311/Endpoint/Integration/SalesForceRest.pm:208]: As the other create_object call is hidden in the integration package, should this be similarly?

File: perllib/Integrations/SalesForceRest.pm [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+            "Error fetching from SalesForce: %s (%s)\n",
+            $message,
+            $code
+        );
+    }
+
+    return $content;
+}
+
+sub get_requests {
```
Comment (dracos) [inline perllib/Integrations/SalesForceRest.pm:149]: Some unused functions in here.

File: perllib/Open311/Endpoint/Integration/SalesForceRest.pm [hunk 4]
```diff
@@ -0,0 +1,214 @@
+package Open311::Endpoint::Integration::SalesForceRest;
+
+use Moo;
+extends 'Open311::Endpoint';
+with 'Open311::Endpoint::Role::mySociety';
+
+use Open311::Endpoint::Service::Request::SalesForce;
+use Open311::Endpoint::Service::Attribute;
+use Open311::Endpoint::Service::UKCouncil;
```
[the remark this answers is not in the document]
Comment (struan) [reply perllib/Open311/Endpoint/Integration/SalesForceRest.pm]: Yes, I will squash it all down post this review

File: perllib/Open311/Endpoint/Integration/SalesForceRest.pm [hunk 5]
```diff
@@ -100,17 +100,17 @@ sub services {
...
 
         next if $self->blacklist->{$service->{value}};
 
-        my $service = Open311::Endpoint::Service::UKCouncil->new(
+        my $service = Open311::Endpoint::Service::UKCouncil::Salesforce->new(
             service_name => $service->{label},
             description => $service->{label},
             service_code => $service->{value},
-            $service->{group} ? ( group => $service->{group} ) : (),
+            $service->{groups} ? ( groups => $service->{groups} ) : (),
```
Comment (dracos) [inline perllib/Open311/Endpoint/Integration/SalesForceRest.pm:114]: Should this be checking @{$service->{groups}} as well (and meta down below)? Would it be better if the integration guaranteed to return an arrayref, then this code wouldn't need to do two checks whenever?

File: perllib/Integrations/SalesForceRest.pm [hunk 6]
```diff
@@ -243,14 +245,23 @@ sub search {
...
 sub _get_pos {
     my $s = shift;
 
     my $bytes = decode_base64($s);
     my $len = length($bytes) * 8;
     my $pos_map = unpack("B$len", $bytes);
-    $pos_map =~ s/^(0*1).*$/$1/;
+    my @pos;
+    my $pos = 0;
+    foreach (split /1/i, $pos_map) {
```
Comment (dracos) [inline perllib/Integrations/SalesForceRest.pm:262]: This works, but as it needs to traverse the entire string anyway to do the split, would it be easier if it were:
`for (my $pos=0; $pos<length $pos_map; $pos++) { push @pos, $pos if substr($pos_map, $pos, 1) == 1; }`

**Comment by struan** [comment]
This was merged ages ago

**Pull request closed without merging**